src/Npgsql/ConnectorPool.cs

```diff
@@ -197,200 +175,188 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
-                    CheckInvariants(newState);
-                    if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) != state.All)
+                    // We're under the pool's max capacity, "allocate" a slot for a new physical connection.
+                    // Don't spin for this https://github.com/dotnet/coreclr/pull/21437
+                    while (true)
                     {
-                        // Our attempt to increment the busy count failed, Loop again and retry.
-                        continue;
+                        var openCountPrev = Interlocked.CompareExchange(ref State.Open, openCount + 1, openCount);
+                        if (openCountPrev == openCount) break;
+
+                        openCount = openCountPrev;
                     }
 
                     try
                     {
-                        // We've managed to increase the busy counter, open a physical connections
+                        // We've managed to increase the open counter, open a physical connections.
                         connector = new NpgsqlConnector(conn) { ClearCounter = _clearCounter };
                         await connector.Open(timeout, async, cancellationToken);
-                    }
-                    catch
-                    {
-                        // Physical open failed, decrement busy back down
-                        conn.Connector = null;
 
-                        var sw = new SpinWait();
-                        while (true)
+                        // We immediately store the connector as well, assigning it an index
+                        // that will be used during the lifetime of the connector for both _idle and _open.
+                        for (var i = 0; i < _open.Length; i++)
                         {
-                            state = State.Copy();
-                            newState = state;
-                            newState.Busy--;
-                            if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) != state.All)
+                            if (Interlocked.CompareExchange(ref _open[i], connector, null) == null)
                             {
-                                // Our attempt to increment the busy count failed, Loop again and retry.
-                                sw.SpinOnce();
-                                continue;
+                                connector.PoolIndex = i;
+                                break;
                             }
-
-                            break;
                         }
 
-                        // There may be waiters because we raised the busy count (and failed). Release one
-                        // waiter if there is one.
-                        if (_waiting.TryDequeue(out var waitingOpenAttempt))
+                        // Start the pruning timer if we're above MinPoolSize
+                        if (_pruningTimer == null && openCount > _min)
```

Comment:
Had some historical reasons (if I remember correctly because of changes to code flow for TCS and Task.Yield)
Will examine

Reply:
Right, so I'm not entirely relaxed with keeping all the interactions with the prune timer and perf counters outside of this critical try block. 
If we fail to return the connector because of an exception in either of those we've effectively lost a slot and an actual active connection for the lifetime of the pool. 

Timers interact with the timer queue, not sure if it never throws. 
But more so the Diagnostic counters are an abstraction over perf counters and probably soon also over event counters. I'm not sure if either of those guarantee to never throw and rather not rely on it either. Best to play it safe here, the decrement is important.

Reply:
I'd definitely wouldn't want an exception from the timer or any perf counters to be handled in the same way as we handle connector creation failure... Rather than failing the open (as we must do when the connector can't be opened), for these kinds of (totally theoretical!) exceptions it would be better to log and ignore... So if I were worried about that I'd have a second try/catch doing something else anyway.

But I don't think that really makes sense either. We have various counter code all around the pool (as well as various other code), without any try/catch - and we've never had any sort of reports of exceptions (which would bubble up to the user).

I understand the idea of being defensive, but I'd rather be sure I know exactly which exceptions can be generated where, and handle them appropriately - rather than blanket try/catch for theoretical exceptions that may occur.

Reply:
BTW the new event counter support would probably be additional calls (in very similar places) rather then within the current counter API, which would remain only for .NET Framework (for .NET Core it's totally empty methods which get elided). Those new methods would effectively be an Interlocked.Add() operation on some static counter, so again I wouldn't put them in a try/catch.

Reply:
>So if I were worried about that I'd have a second try/catch doing something else anyway.

I'm fine with doing that too, all I want is to make sure the pool can't get into a bad state (that would also mean adding a null check to Release btw, as I found out during the double Close bug)

>BTW the new event counter support would probably be additional calls (in very similar places) rather then within the current counter API, which would remain only for .NET Framework (for .NET Core it's totally empty methods which get elided). Those new methods would effectively be an Interlocked.Add() operation on some static counter, so again I wouldn't put them in a try/catch.

Why would you not ifdef that under the diagnostic counter abstraction that's already there?

Reply:
> all I want is to make sure the pool can't get into a bad state (that would also mean adding a null check to Release btw, as I found out during the double Close bug)

So once again, if you really want to go down this path, you should probably wrap almost all lines in the pool code with try/catch... I'd rather not put try/catch around lines where exceptions are not supposed to happen - and if they do, we can and should deal with them at that point... But if you feel very strongly about this, and you think this specific code fragment is somehow special and merits special guarding against exceptions, I don't mind that much.

> Why would you not ifdef that under the diagnostic counter abstraction that's already there?

The current counter abstraction was done to support the old Windows Perf Counters, and it's not very suitable for the new event counters... The abstraction has different instances of the Counter type, each of which works has a string identifier - because that's how the old perf counters worked (classical example of a premature/leaky abstraction).

With the new event counters all we want is to do an interlocked operation on some static counter somewhere - there's simply no use for that abstraction IMHO./